torch/_subclasses/fake_tensor.py

```diff
@@ -417,18 +431,17 @@ def nyi(fake_mode, func, *args, **kwargs):
 @contextlib.contextmanager
 def in_kernel_invocation_manager(fake_mode):
     # See: note [Fake Tensor Dispatch Keys]
+    prev_in_kernel = fake_mode.in_kernel_invocation
     meta_in_tls = torch._C._meta_in_tls_dispatch_include()
-    prev = fake_mode.in_kernel_invocation
+    assert meta_in_tls == prev_in_kernel, f"{meta_in_tls}, {prev_in_kernel}"
```

Comment:
If these two pieces of state have to always be synchronized, why not just do away with `fake_mode.in_kernel_invocation` and only use the TLS instead?

Reply:
This has already proven useful in debugging multi-threaded inference where we got into a weird state. It is possible for us to get into a weird state in the future with multiple fake tensor modes or something along those lines. Would prefer the separate properties and assertion.

Reply:
What was the resolution for multithreaded inference? If the answer was "we can't do a global state and need TLS", only using the TLS would have prevented the problem in the first place, no?

Reply:
Actually, IIUC, you resolved the problem by disabling multithreading, is that right? But actually, that's papering over the root cause of this problem, which is that in kernel invocation cannot be a per Fake Tensor object property!

Reply:
Like, I think disabling multithreading is still a valid thing to do, but you should still fix the root cause of the bug.

Reply:
> which is that in kernel invocation cannot be a per Fake Tensor object property
It's a property on the FakeTensorMode, not on the fake tensor. The problem was concurrent execution of `__torch_dispatch__` where one thread was in kernel invocation so the other thread reported "meta" when it shouldn't have.   

> only using the TLS would have prevented the problem in the first place, no?

Hmm maybe, but there are other potential data races, e.g. in the MetaConverter for preserving aliasing. I would rather not go into the rabbit hole of trying to make Fake Tensor Mode thread safe, especially since it's going to be really hard to test and validate. 

I think the right fix is to have a lock on FakeTensorMode only allowing one thread at a time invoke `__torch_dispatch__`.

Reply:
> It's a property on the FakeTensorMode, not on the fake tensor. The problem was concurrent execution of __torch_dispatch__ where one thread was in kernel invocation so the other thread reported "meta" when it shouldn't have.

Sorry, I wasn't clear. You should have a *per mode* TLS. This can be done by allocating a threading.local per fake tensor

> Hmm maybe, but there are other potential data races, e.g. in the MetaConverter for preserving aliasing. I would rather not go into the rabbit hole of trying to make Fake Tensor Mode thread safe, especially since it's going to be really hard to test and validate.

You cannot have a true data race, because fake tensor is written in Python and for Python code to execute the GIL must be held. You can only have "logical" races, where you do some change and then release the GIL (by calling into PyTorch) allowing another thread to observe the problem. 

A lock around the entire shebang will certainly work, but if you have the lock have a large extent you have to be more worried about accidentally deadlocking (in particular, if you `with self` inside the fake tensor mode, you had better release the lock). I'd prefer a smaller extent lock; e.g., lock FakeConverter, and use TLS for in kernel invocation.

Reply:
> You can only have "logical" races, where you do some change and then release the GIL (by calling into PyTorch) allowing another thread to observe the problem.

Yea, what I meant was two threads concurrently check `get_tensor_memo`, and return two separate storages, as opposed to data corruption or something.

> in particular, if you with self inside the fake tensor mode, you had better release the lock

The idea was to only allow one thread access at a time, so this wouldn't be an issue. But a FakeConverter lock and TLS for in kernel invocation could also work.

I agree would be good to fix multi-threading, but at least for PT2, not a critical issue, so will do in follow up.